src/L2OALM.jl

```diff
@@ -1,5 +1,426 @@
+"""
+function _pre_hook_primal(
+    θ,
+    primal_model,
+    train_state_primal,
+    dual_model,
+    train_state_dual,
+    bm,
+)
+    # Forward pass for dual model
+    dual_hat_k, _ = dual_model(θ, train_state_dual.parameters, train_state_dual.states)
+
+    return (dual_hat_k,)
+end
+
+"""
+    _pre_hook_dual(θ, primal_model, train_state_primal, dual_model, train_state_dual, bm)
+
+Default pre-hook function for the dual model in the L2O-ALM algorithm.
+This function performs a forward pass through the primal model to obtain the predicted state and constraints.
+"""
+function _pre_hook_dual(
+    θ,
+    primal_model,
+    train_state_primal,
+    dual_model,
+    train_state_dual,
+    bm,
+)
+    # # Forward pass for primal model
+    X̂, _ = primal_model(θ, train_state_primal.parameters, train_state_primal.states)
+    gh = constraints!(bm, X̂, Θ)
+
+    # Forward pass for dual model
+    dual_hat, _ = dual_model(θ, train_state_dual.parameters, train_state_dual.states)
+
+    return (dual_hat, gh)
+end
+
+"""
+    _reconcile_alm_primal_state(batch_states::Vector{NamedTuple})
+
+Default function that reconciles the state of the primal model after processing a batch of data.
+This function computes the maximum violation, mean violations, mean objectives, and total loss
+from the batch states.
+"""
+function _reconcile_alm_primal_state(batch_states::Vector{NamedTuple})
+    max_violation = maximum([s.new_max_violation for s in batch_states])
+    mean_violations = mean([s.mean_violations for s in batch_states])
+    mean_objs = mean([s.mean_objs for s in batch_states])
+    mean_loss = mean([s.total_loss for s in batch_states])
+    return (;
+        new_max_violation = max_violation,
+        mean_violations = mean_violations,
+        mean_objs = mean_objs,
+        total_loss = mean_loss,
+    )
+end
+
+"""
+    _reconcile_alm_dual_state(batch_states::Vector{NamedTuple})
+
+Default function that reconciles the state of the dual model after processing a batch of data.
+This function computes the mean dual loss from the batch states.
+"""
+function _reconcile_alm_dual_state(batch_states::Vector{NamedTuple})
+    dual_loss = mean([s.dual_loss for s in batch_states])
+    return (dual_loss = dual_loss,)
+end
+
+"""
+    _update_ALM_ρ!(hpm::Dict{Symbol, Any}, current_state::NamedTuple)
+
+Default function to update the hyperparameter ρ in the ALM algorithm.
+This function increases ρ by a factor of α if the new maximum violation exceeds τ times the previous maximum violation.
+"""
+function _update_ALM_ρ!(
+    hpm_primal::Dict{Symbol,Any},
+    hpm_dual::Dict{Symbol,Any},
+    current_state::NamedTuple,
+)
+    if current_state.new_max_violation > hpm_primal.τ * hpm_primal.max_violation
+        hpm_primal[:ρ] = min(hpm_primal[:ρmax], hpm_primal[:ρ] * hpm_primal[:α])
+        hpm_dual[:ρ] = hpm_primal[:ρ]  # Ensure dual model uses the same ρ
+    end
+    hpm_primal.max_violation = current_state.new_max_violation
```

Comment:
```suggestion
    if current_state.max_violation > hpm_primal.τ * hpm_primal.max_violation
        hpm_primal[:ρ] = min(hpm_primal[:ρmax], hpm_primal[:ρ] * hpm_primal[:α])
        hpm_dual[:ρ] = hpm_primal[:ρ]  # Ensure dual model uses the same ρ
    end
    hpm_primal.max_violation = current_state.max_violation
```

https://github.com/LearningToOptimize/L2OALM.jl/pull/2/files#r2224260241 cont.